### PR TITLE
feat(yarn): optimize yarn binary when binarySource=docker

### LINKED ISSUE #8723: Yarn optimization command does not handle pinned yarn binaries
**What Renovate type, platform and version are you using?**

Hosted + Github

**Describe the bug**

When using [yarn policies](https://classic.yarnpkg.com/en/docs/cli/policies/#toc-policies-set-version), the yarn binary is checked out as part of the repo and its path is written to `.yarnrc`:

```
# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


yarn-path ".yarn/releases/yarn-1.22.10.cjs"
```

Renovate's [optimize command](https://github.com/renovatebot/renovate/blob/master/lib/manager/npm/post-update/yarn.ts#L38-L39) has no effect because it patches the wrong yarn binary (global yarn instead of local yarn). This causes `yarn install` to either timeout (due to slow writes from container to host), or run out of disk space.

**Relevant debug logs**

https://gist.github.com/ldiqual/a9c1caf1a6839f6579f4abcefba34229

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/master/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] This is a really small bug, it does not need a reproduction (think small typo)
- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

A couple ideas on how to fix this:
 - Patch both global yarn and local yarn by following path in `yarnrc`
 - Don't do a volume mount, instead copy the repo in docker container, run yarn install in there (which will have much faster node_modules writes), then copy `yarn.lock` back into repo on host.
 - Use npm to generate yarn lock: https://stackoverflow.com/a/58609139/646960

**Current workaround**

Add `"yarnrc": ""` to renovate.json so that it overrides the local `.yarnrc`, skipping local yarn entirely and using the global yarn which is correctly patched.

### PR DESCRIPTION
**Pull request by rarkins**
## Changes:

Optimizes the `yarn-path` binary if found when in docker mode.

## Context:

Without doing this, all packages will be downloaded and installed. Closes #8723 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [x] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/npm/post-update/yarn.ts [hunk 1]
```diff
@@ -21,25 +21,36 @@ export interface GenerateLockFileResult {
...
         .split('\n')
         .find((line) => line.startsWith('yarn-offline-mirror '));
-      if (mirrorLine) {
-        return true;
+      offlineMirror = !!mirrorLine;
+      const pathLine = yarnrc
+        .split('\n')
+        .find((line) => line.startsWith('yarn-path '));
+      if (pathLine) {
+        yarnPath = pathLine.replace('yarn-path ', '');
```
Comment (viceice) [inline lib/manager/npm/post-update/yarn.ts:40]: I think `yarn-path` ca be quoted

**Review by viceice — CHANGES_REQUESTED** [review]
Needs coverage

File: lib/manager/npm/post-update/yarn.ts (hunk 1 above)
Comment (rarkins) [reply lib/manager/npm/post-update/yarn.ts]: Thanks, now addressed in latest commit

File: lib/manager/npm/post-update/yarn.ts [hunk 2]
```diff
@@ -21,25 +21,36 @@ export interface GenerateLockFileResult {
...
         .split('\n')
         .find((line) => line.startsWith('yarn-offline-mirror '));
-      if (mirrorLine) {
-        return true;
+      offlineMirror = !!mirrorLine;
+      const pathLine = yarnrc
+        .split('\n')
+        .find((line) => line.startsWith('yarn-path '));
+      if (pathLine) {
+        yarnPath = pathLine.replace(/^yarn-path\s+"?(.*?)"?$/, '$1');
```
Comment (viceice) [inline lib/manager/npm/post-update/yarn.ts:40]:
```suggestion
        yarnPath = pathLine.replace(/^yarn-path\s+"?(.+?)"?$/, '$1');
```

Should not empty.

**Review by viceice — APPROVED** [review]

**Pull request merged**